packages/salesforcedx-vscode-core/src/commands/baseDeployRetrieve.ts

```diff
@@ -104,22 +123,103 @@ export abstract class DeployRetrieveExecutor<
   protected abstract postOperation(
     result: DeployRetrieveResult | undefined
   ): Promise<void>;
+  protected abstract handleSourceConflictError(e: any): void;
 }
 
 export abstract class DeployExecutor<T> extends DeployRetrieveExecutor<T> {
   protected async doOperation(
     components: ComponentSet,
     token: vscode.CancellationToken
   ): Promise<DeployResult | undefined> {
+    const projectPath = getRootWorkspacePath();
+    const connection = await WorkspaceContext.getInstance().getConnection();
+
+    const sourceTracking = await SourceTrackingService.createSourceTracking(
+      projectPath,
+      connection
+    );
+    await sourceTracking.ensureLocalTracking();
+
     const operation = await components.deploy({
-      usernameOrConnection: await WorkspaceContext.getInstance().getConnection()
+      usernameOrConnection: connection
     });
 
     this.setupCancellation(operation, token);
 
     return operation.pollStatus();
   }
 
+  protected async handleSourceConflictError(e: any) {
+    const componentPaths = e.data.map(
+      (component: { filePath: any }) => component.filePath
+    );
+    const username = await ConfigUtil.getUsername();
+    const metadataCacheService = new MetadataCacheService(String(username));
+    const cacheResult = await metadataCacheService.loadCache(
+      componentPaths,
+      workspaceUtils.getRootWorkspacePath(),
+      false
+    );
+
+    const detector = new TimestampConflictDetector();
+    const diffs = detector.createDiffs(cacheResult, true);
+
+    const conflictMessages = this.getMessagesFor(this.logName);
```

Comment:
Are `conflictMessages` and the following conditional being tested for? I found that block code a little tricky to read, so it'd be nice to have some additional coverage there.

Reply:
Yep.  I refactored these out into a separate file since they are now shared between multiple modules.  Open to any ideas for better way of doing this!

Reply:
I'll take a look re: test coverage.

Reply:
I added a test file to unit test this messages module.  Let me know if you have any feedback on it!